agent/eventhandler/task_handler.go

```diff
@@ -218,40 +230,47 @@ func (handler *TaskHandler) taskStateChangesToSend() []api.TaskStateChange {
 			events = append(events, event)
 		}
 	}
-	return events
-}
-
-// batchContainerEventUnsafe collects container state change events for a given task arn
-func (handler *TaskHandler) batchContainerEventUnsafe(event api.ContainerStateChange) {
-	// if this is a managed agent state change, let's try to bundle it in an existing container event
-	if event.GetEventType() == statechange.ManagedAgentEvent {
-		containerChanges, ok := handler.tasksToContainerStates[event.TaskArn]
-		if ok {
-			for i, cs := range containerChanges {
-				if cs.ContainerName == event.ContainerName && cs.Status == event.Status {
-					cs.ManagedAgents = append(cs.ManagedAgents, event.ManagedAgents...)
-					containerChanges[i] = cs
-					return
-				}
+	for taskARN := range handler.tasksToManagedAgentStates {
+		for _, event := range events {
+			if event.TaskARN == taskARN {
+				continue
 			}
 		}
-	}
+		if task, ok := handler.state.TaskByArn(taskARN); ok {
+			// We do not allow the ticker to submit managed agent state updates for
+			// tasks that are STOPPED. This prevents the ticker's asynchronous
+			// updates from clobbering managed agent states when the task
+			// transitions to STOPPED, since ECS does not allow updates to
+			// managed agent states once the task has moved to STOPPED.
+			knownStatus := task.GetKnownStatus()
+			if knownStatus >= apitaskstatus.TaskStopped {
+				continue
+			}
+			event := api.TaskStateChange{
+				TaskARN: taskARN,
+				Status:  task.GetKnownStatus(),
+				Task:    task,
+			}
+			event.SetTaskTimestamps()
 
-	// For now, evey time a container stops, we also send stop for exec command agent
-	if event.GetEventType() == statechange.ContainerEvent && event.Status == status.ContainerStopped {
-		if _, ok := event.Container.GetManagedAgentByName(execcmd.ExecuteCommandAgentName); ok {
-			event.ManagedAgents = []api.ManagedAgentStateChange{{
-				Name:   execcmd.ExecuteCommandAgentName,
-				Status: status.ManagedAgentStopped,
-				Reason: "Parent container stopped",
-			}}
+			events = append(events, event)
```

Comment:
I believe this is not going to work. We might need to dedupe `TaskStateChange`s in this method

Reply:
dedupe logic is already present, modified that a bit.